From 2152fbed8d14aefbff20dd18203689b810dbc316 Mon Sep 17 00:00:00 2001
From: Carlos Bentzen <cadubentzen@igalia.com>
Date: Mon, 13 Jan 2025 15:55:26 +0100
Subject: [PATCH] webrtc: fix duplicate payload types with RTX and multiple
 video codecs

Before this patch, there could be duplicate payload types in offers that
have, within a media section, multiple codecs and RTX enabled:

```
m=video 9 UDP/TLS/RTP/SAVPF 96 97 97 <-- HAS DUPLICATES
a=sendrecv
a=rtpmap:96 VP8/90000
a=rtcp-fb:96 nack
a=rtcp-fb:96 nack pli
a=rtcp-fb:96 ccm fir
a=rtcp-fb:96 transport-cc
a=rtpmap:97 H264/90000
a=rtcp-fb:97 nack
a=rtcp-fb:97 nack pli
a=rtcp-fb:97 ccm fir
a=rtcp-fb:97 transport-cc
a=rtpmap:97 rtx/90000  <--------- PT IS DUPLICATE
a=fmtp:97 apt=96
```

Fix this by populating the media_mapping array with all media formats
rather than only the first one. The added test case reproduces the issue,
which fails without this patch.

Part-of: <https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/8259>
---
 ext/webrtc/gstwebrtcbin.c | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/ext/webrtc/gstwebrtcbin.c b/ext/webrtc/gstwebrtcbin.c
index d26c814d2..b9f1727b2 100644
--- a/ext/webrtc/gstwebrtcbin.c
+++ b/ext/webrtc/gstwebrtcbin.c
@@ -3581,6 +3581,7 @@ sdp_media_from_transceiver (GstWebRTCBin * webrtc, GstSDPMedia * media,
   for (i = 0; i < gst_caps_get_size (caps); i++) {
     GstCaps *format = gst_caps_new_empty ();
     GstStructure *s = gst_structure_copy (gst_caps_get_structure (caps, i));
+    guint media_pt;
 
     if (i == 0) {
       gst_structure_foreach (extmap, (GstStructureForeachFunc) _copy_field, s);
@@ -3603,6 +3604,9 @@ sdp_media_from_transceiver (GstWebRTCBin * webrtc, GstSDPMedia * media,
       return FALSE;
     }
 
+    if (gst_structure_get_int (s, "payload", &media_pt))
+      find_or_create_payload_map_for_media_pt (media_mapping, media_pt);
+
     gst_caps_unref (format);
   }
 
@@ -3615,10 +3619,7 @@ sdp_media_from_transceiver (GstWebRTCBin * webrtc, GstSDPMedia * media,
     guint rtx_target_ssrc = -1;
     gint media_pt;
 
-    if (gst_structure_get_int (s, "payload", &media_pt) &&
-        webrtc->bundle_policy == GST_WEBRTC_BUNDLE_POLICY_NONE)
-      find_or_create_payload_map_for_media_pt (media_mapping, media_pt);
-
+    gst_structure_get_int (s, "payload", &media_pt);
     rtx_target_pt = media_pt;
 
     if (!gst_structure_get_int (s, "clock-rate", &clockrate))
-- 
2.47.1

